src/hotspot/share/nmt/memLogRecorder.cpp

```diff
@@ -0,0 +1,872 @@
+              MemTracker::record_virtual_memory_commit(e->ptr, e->size, stack);
+              //fprintf(stderr, "]\n");fflush(stderr);
+              break;
+            case NMT_VirtualMemoryLogRecorder::Type::SPLIT_RESERVED:
+              //fprintf(stderr, "[MemTracker::record_virtual_memory_split_reserved\n");
+              MemTracker::record_virtual_memory_split_reserved(e->ptr, e->size, e->size_split, mem_tag, NMTUtil::index_to_tag((int)e->mem_tag_split));
+              //fprintf(stderr, "]\n");fflush(stderr);
+              break;
+            case NMT_VirtualMemoryLogRecorder::Type::TAG:
+              //fprintf(stderr, "[record_virtual_memory_type(%p, %zu, %p)\n", e->ptr, e->size, &stack);fflush(stderr);
+              MemTracker::record_virtual_memory_tag(e->ptr, mem_tag);
+              //fprintf(stderr, "]\n");fflush(stderr);
+              break;
+            default:
+              fprintf(stderr, "HUH?\n");
+              os::exit(-1);
+              break;
+          }
+        }
+        jlong end = os::javaTimeNanos();
+        jlong duration = (start > 0) ? (end - start) : 0;
+        total += duration;
+      }
+    }
+    fprintf(stderr, "samples:%ld total:%ld\n", count, total);
+
+    //sleep(1);h
+
+    _close_and_check(records_fi.fd);
+
+#if defined(LINUX) || defined(__APPLE__)
+    pthread_mutex_unlock(&VirtualMemoryLogInfo._mutex);
+#elif defined(WINDOWS)
+    // ???
+#endif
+
+    os::exit(0);
+  }
+}
+
+static inline const char* type_to_name(NMT_VirtualMemoryLogRecorder::Type type) {
+  switch (type) {
+    case NMT_VirtualMemoryLogRecorder::Type::RESERVE: return "RESERVE"; break;
+    case NMT_VirtualMemoryLogRecorder::Type::RELEASE: return "RELEASE"; break;
+    case NMT_VirtualMemoryLogRecorder::Type::UNCOMMIT: return "UNCOMMIT"; break;
+    case NMT_VirtualMemoryLogRecorder::Type::RESERVE_AND_COMMIT: return "RESERVE_AND_COMMIT"; break;
+    case NMT_VirtualMemoryLogRecorder::Type::COMMIT: return "COMMIT"; break;
+    case NMT_VirtualMemoryLogRecorder::Type::SPLIT_RESERVED: return "SPLIT_RESERVED"; break;
+    case NMT_VirtualMemoryLogRecorder::Type::TAG: return "TAG"; break;
+    default: break;
+  }
+}
+
+void NMT_VirtualMemoryLogRecorder::log(NMT_VirtualMemoryLogRecorder::Type type, MemTag mem_tag, MemTag mem_tag_split, size_t size, size_t size_split, address ptr, const NativeCallStack *stack) {
+  //fprintf(stderr, "NMT_VirtualMemoryLogRecorder::log (%s, %hhu, %hhu, %zu, %zu, %p, %p)\n",
+  //        type_to_name(type), mem_tag, mem_tag_split, size, size_split, ptr, stack);fflush(stderr);
+  if (VirtualMemoryLogInfo._done) {
+    return;
+  }
+  volatile static jlong _count = 0;
+  jlong count = -1;
+
+  if (VirtualMemoryLogInfo._initiliazed && !VirtualMemoryLogInfo._done) {
+#if defined(LINUX) || defined(__APPLE__)
+    pthread_mutex_lock(&VirtualMemoryLogInfo._mutex);
+#elif defined(WINDOWS)
```

Comment:
Won't this code always be run under the NMT lock anyway? Could just use that, especially as NMT now has a unique global lock.